Feature/board - The Board application is added

elliotyim opened the pull request:
I added a Board application and need a customised user model.
As you could find from my commit messages,
A Photo application is also needed for both the User application and the Board application.
If there are not any problem, I would like to start creating features of basic board which don't have photos yet.

I also wonder if I should delete views.py and tests.py in the Board application
because both of them do not seem to be required.

Let me know anytime if I am doing something wrong.

elliotyim committed:
Add some files in .idea to prevent any conflicts

elliotyim committed:
Initialise an app: board

elliotyim committed:
Add models and register models to admin

In terms of extention, Photo model should not be applied to Board app
It needs to be added into Photo app so other app can use it.

Here are the models that would be contained for now
: Category, Board, Comment, Tag

In addition, the User app should contains customised User model and its manager.

elliotyim committed:
Modify user rows in models

Basic User model is applied to make Mypy of Jenkins pass the models

elliotyim committed:
Fix models

Imported and designated the User model directly from Django to prevent Mypy error.

hyeonjames commented on .gitignore:
```diff
@@ -266,4 +266,12 @@ fabric.properties
 
-# End of https://www.gitignore.io/api/django,python,pycharm,visualstudiocode
\ No newline at end of file
+# End of https://www.gitignore.io/api/django,python,pycharm,visualstudiocode
```
이거 맨 끝줄에다가 넣어주시고, 추가한 부분에 대해서는 코멘트 정도만 해주세요

elliotyim replied:
제일 밑으로 옮기고 주석으로 pycharm 관련 세팅파일들이라고 적었습니다.

hyeonjames commented on board/admin.py:
```diff
@@ -0,0 +1,9 @@
+from django.contrib import admin  # noqa: 401
```
noqa는 필요 없죠, 밑에서 사용하고 있으니까 오류 안날꺼에요. 빼주세요

elliotyim replied:
넵

hyeonjames commented on board/models.py:
```diff
@@ -0,0 +1,64 @@
+from django.contrib.auth.models import User
+from django.db import models  # noqa: F401
```
마찬가지로 밑에서 사용하고 있으니 noqa는 필요가 없어요. 빼주세여

elliotyim replied:
넵

hyeonjames commented on board/models.py:
```diff
@@ -0,0 +1,64 @@
+
+    title = models.CharField(max_length=30, db_index=True)
+    content = models.TextField()
+    meeting_location = models.CharField(max_length=75)
```
앞에 접두사 `meeting`은 필요가 없어요. 그냥 `location`, `capacity`, `date` 등으로 만들어주세요.
(단순히 meeting 을 위한 게 아니니까요)

hyeonjames replied:
아, 추가로 주소가 75자는 너무 짧은 듯요? 적당히 255자로 하는게 좋겠어요.
마찬가지로 `title` 같은 것도 너무 짧아요.

elliotyim replied:
접두사 전부 빼고 주소는 255자로, title은 50자로 늘렸습니다.

hyeonjames commented on board/models.py:
```diff
@@ -0,0 +1,64 @@
+    """
+    title = models.CharField(max_length=30)
+
+    board_tags = models.ManyToManyField('Board')
```
잠깐 이게 왜 여기 있죠? 반대인거 같은데요, 게시글 쪽에 `tags`를 가지고 있어야 되는 형태여야 될거 같아요

elliotyim replied:
board_tags의 이름을 post_tags로 바꾸고 Post 모델쪽으로 이동시켰습니다.

hyeonjames commented on board/models.py:
```diff
@@ -0,0 +1,64 @@
+    meeting_times_of_day = models.IntegerField()
+    created_date = models.DateTimeField(auto_now_add=True)
+
+    author = models.ForeignKey(User, on_delete=models.CASCADE)
```
모델 속성들을 보고나니까 이건 `Board`라고 볼 수가 없네요. 이름을 바꾸셔야 될거 같아요. 엄청 헷갈리네요
`Board`라는 건 "게시판"을 의미하는 거고, 이 속성들은 "게시글" 이라고 봐야 될거 같은데요..

elliotyim replied:
넵 Post로 바꿨습니다.

hyeonjames commented on board/models.py:
```diff
@@ -0,0 +1,64 @@
+
+    board = models.ForeignKey('Board', on_delete=models.CASCADE)
+    parent_comment = models.ForeignKey('self', on_delete=models.SET_NULL, null=True)
+    user = models.ForeignKey(User, on_delete=models.CASCADE)
```
`user`보다는 `author`가 어울리겠네요

hyeonjames replied:
`on_delete`를 `CASCADE`로 잡는 것은 잘 생각해 봐야 되는 문제에요. 이 뜻은 사용자가 삭제되면 코멘트도 같이 삭제된다는 뜻인데, 코멘트 같은 것들은 삭제 안하는게 낫습니다. 기록에 남지가 않아서 추적하기가 힘들어요

elliotyim replied:
author로 바꾸고 on_delete를 전부 PROTECT로 바꿨습니다.
그리고 중요 모델들에 대해서는 삭제가 아닌 비활성화로 바꾸기 위해 is_active 필드를 추가하고 디폴트 값을 False로 바꿨습니다.

hyeonjames commented on board/models.py:
```diff
@@ -0,0 +1,64 @@
+    created_date = models.DateTimeField(auto_now_add=True)
+
+    board = models.ForeignKey('Board', on_delete=models.CASCADE)
+    parent_comment = models.ForeignKey('self', on_delete=models.SET_NULL, null=True)
```
이런 문제가 있어요.
A라는 사람이 코멘트를 달고 B 라는 사람이 그 코멘트에 답글을 달았어요.
A라는 사람이 코멘트를 삭제 했어요. 그럼 B라는 사람이 남긴 코멘트가 parent가 NULL이 되면서, 그 답글은 답글이 아닌 그냥 댓글로 인식이 될거에요

elliotyim replied:
이를 방지 하기 위해 코멘트의 삭제를 못하게 PROTECT로 바꿨습니다

hyeonjames commented on board/models.py:
```diff
@@ -0,0 +1,64 @@
+    created_date = models.DateTimeField(auto_now_add=True)
+
+    author = models.ForeignKey(User, on_delete=models.CASCADE)
+    category = models.ForeignKey('Category', on_delete=models.CASCADE)
```
`CASCADE` 노노, 이거 역시 카테고리가 사라졌다고 게시글이 사라지는 건 별로 좋지가 않아요.;; 기록상 남기는 게 좋아요.

elliotyim replied:
넵

hyeonjames commented on board/models.py:
```diff
@@ -0,0 +1,64 @@
+    Board model which has simple information.
+    """
+
+    title = models.CharField(max_length=30, db_index=True)
```
title처럼 문자열이고 꽤나 긴 형태로 들어가는 경우는 인덱스를 만드는게 좋은 생각인지 모르겠어요.
 `db_index` 는 추후에 생각해보고 넣는걸로 하죠.

elliotyim replied:
넵 db_index를 제거했습니다.

elliotyim committed:
Remove unneccessary files

elliotyim committed:
Rename the applicaion and modify models

- Renamed the application and models: Board -> Post
- Additional infomation for docs of the Post model: meeting_time_of_day
- Added post status & comments status (is_active) for deletion.
- Added lookups to the Category & the Tag models: unique=True

elliotyim committed:
Add description into .gitignore

- Marked end of the line.
- Added proper description for what are added into .gitignore.

elliotyim committed:
Remove the unnecessary comment

elliotyim committed:
Fix Post model

- Removed #noaq: F401
- Removed prefixes `meeting` from location, capacity, date
- Increased max length of location(to 255), titles(to 50)
- Moved post_tags to Post model from Tag model.
- Rename user to author.
- Make all `on_delete` lookups PROTECT as it will never be deleted actually.
- Removed db_index for good.

hyeonjames commented on post/models.py:
```diff
@@ -0,0 +1,69 @@
+
+    author = models.ForeignKey(User, on_delete=models.PROTECT)
+    category = models.ForeignKey('Category', on_delete=models.PROTECT)
+    post_tags = models.ManyToManyField('Post')
```
그냥 ‘tags’ 가 맞을듯여

elliotyim replied:
넵 tags로 수정하였습니다.

hyeonjames commented on post/models.py:
```diff
@@ -0,0 +1,69 @@
+    post_tags = models.ManyToManyField('Post')
+
+    def __str__(self):
+        return self.title
```
str같은 경우 로그를 위해서 많이 쓰여요. 그런데 title 은 충분히 중복될 수 있으니 title만 넣는 건 좋지 않을 듯 하네요.
`f'#{self.id} : {self.title}'` 이런식으로 로그를 써도 찾아볼 수 있는 내용을 쓰는게 좋겠어요.

elliotyim replied:
post, category, tag 전부 해당 방식으로 수정하였습니다.

hyeonjames commented on post/models.py:
```diff
@@ -0,0 +1,69 @@
+    author = models.ForeignKey(User, on_delete=models.PROTECT)
+
+    def __str__(self):
+        return self.content
```
`Post` 쪽과 마찬가지로 처리해주세요,
단 `Comment`의 경우 내용이 지나치게 길어질 수 있어요.
따라서 Comment id 정도만 넣는게 좋겠네요

elliotyim replied:
넵

hyeonjames reviewed (changes requested):
develop rebase 한번 해주세요~

elliotyim committed:
Add return type to __str__

Added return type str into all the `__str__` functions of models in post app.

elliotyim committed:
Modify returned __str__ values.

All __str__() of models have changed in order to avoid duplicate names and
printing vague messages for logging.

Formats:
- Category -> #`id` : `title`
- Post -> #`id` : `title`
- Comment -> Comment `id`
- Tag -> #`id` : `title`

hyeonjames commented on post/models.py:
```diff
@@ -0,0 +1,69 @@
+    title = models.CharField(max_length=50, unique=True)
+
+    def __str__(self) -> str:
+        return f'#{self.id} : {self.title}'
```
‘Tag’ 랑 ‘Category’ 쪽은 id 표시가 필요없죠. title이 unique인데
그리구 똑같이 타이틀로 줘버리면 어떤게 Post이고 어떤게 Tag인지 모르는데.. 똑같이 만들어 버리시면 구별을 못해요
Comment 처럼 ‘Category’, ‘Tag’, ‘Post’ 쪽도 표시를 넣어주세요

hyeonjames replied:
차라리, 이런식으로 표시하는 게 나을 듯요.
‘Post(id=..,title=‘’)’
Category랑 Tag는 id가 필요 없으니 title만 넣구요

elliotyim replied:
넹 그게 좋을 것 같습니다.

hyeonjames commented on post/models.py:
```diff
@@ -0,0 +1,69 @@
+
+    author = models.ForeignKey(User, on_delete=models.PROTECT)
+    category = models.ForeignKey('Category', on_delete=models.PROTECT)
+    tags = models.ManyToManyField('Post')
```
Post -> Tag

elliotyim replied:
수정했습니다

elliotyim committed:
Fix models' typo and return values of __str__()

Typo: Post -> Tag

__str__ Formats:
- Category -> Category(title="title")
- Post -> Post(id=`id`, title="title")
- Comment -> Comment (id=`id`)
- Tag -> Tag(title="title")

elliotyim committed:
Remove redundant lookup

Field lookup null has a default value: True

elliotyim committed:
Fix an inappropriate lookup setting

Parent comment should have lookup null=True as it might not have any parent comments.

hyeonjames commented on post/models.py:
```diff
@@ -0,0 +1,69 @@
+    is_active = models.BooleanField(default=True)
+    created_date = models.DateTimeField(auto_now_add=True)
+
+    author = models.ForeignKey(User, on_delete=models.PROTECT)
```
‘CASCADE’ 를 쓸때는 잘 생각해봐야 한다고 했지만 다 PROTECT로 박으란 소린 아닌데 ;;
이렇게 하면 무결성이 떨어져요
Comment 의 Parent_comment 말고는 SET_NULL 해도 되요..

elliotyim replied:
데이터 무결성...그렇군요! 수정했습니다.

elliotyim committed:
Fix a few models' foreign key lookups

All the lookups of foreign key has been changed from PROTECT to SET_NULL
because of data integrity.

elliotyim committed:
Add is_active field into the Category model
